### ISSUE TITLE
Kubernetes properties don't show up

### ISSUE THREAD

**Issue by fbricon** [redhat-developer/vscode-quarkus#129]
Following the instructions from https://quarkus.io/guides/kubernetes-resources, I can't get `kubernetes.group` to show up in application.properties completions.

I'm using this pom.xml:

```xml
<?xml version="1.0"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.redhat.developer</groupId>
  <artifactId>tadaaa</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <properties>
    <surefire-plugin.version>2.22.0</surefire-plugin.version>
    <maven.compiler.target>1.8</maven.compiler.target>
    <quarkus.version>0.27.0</quarkus.version>
    <maven.compiler.source>1.8</maven.compiler.source>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    <maven.compiler.parameters>true</maven.compiler.parameters>
    <compiler-plugin.version>3.8.1</compiler-plugin.version>
  </properties>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-bom</artifactId>
        <version>${quarkus.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-junit5</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.rest-assured</groupId>
      <artifactId>rest-assured</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-mongodb-panache</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-health</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-kubernetes</artifactId>
    </dependency>
  </dependencies>
  <build>
    <plugins>
      <plugin>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <version>${quarkus.version}</version>
        <executions>
          <execution>
            <goals>
              <goal>build</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${compiler-plugin.version}</version>
      </plugin>
      <plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>${surefire-plugin.version}</version>
        <configuration>
          <systemProperties>
            <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
          </systemProperties>
        </configuration>
      </plugin>
    </plugins>
  </build>
  <profiles>
    <profile>
      <id>native</id>
      <activation>
        <property>
          <name>native</name>
        </property>
      </activation>
      <build>
        <plugins>
          <plugin>
            <artifactId>maven-failsafe-plugin</artifactId>
            <version>${surefire-plugin.version}</version>
            <executions>
              <execution>
                <goals>
                  <goal>integration-test</goal>
                  <goal>verify</goal>
                </goals>
                <configuration>
                  <systemProperties>
                    <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
                  </systemProperties>
                </configuration>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
      <properties>
        <quarkus.package.type>native</quarkus.package.type>
      </properties>
    </profile>
  </profiles>
</project>

```

**Comment by angelozerr** [comment]
The https://quarkus.io/guides/kubernetes-resources, speak about `KubernetesConfig` class which should host kubernetes properties.

This class exists in 0.26 -> https://github.com/quarkusio/quarkus/blob/0.26/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesConfig.java

but it seems that it was removed in master, why?

**Comment by fbricon** [comment]
Let's ask @geoand :-)

**Comment by geoand** [comment]
The kubernetes extension was totally overhauled.

Basically `KubernetesConfig` was very limiting so it was removed and the extensions does it's own custom configuration mentioned [here](https://quarkus.io/guides/kubernetes-resources#tuning-the-generated-resources-using-application-properties).

**Comment by fbricon** [comment]
@geoand then the question is, what should we be looking for when inspecting the kubernetes extension jar, in order to identify quarkus properties? Specific annotations?

**Comment by angelozerr** [comment]
Thanks @geoand for your feedback. At first I suggest you to update the doc https://quarkus.io/guides/kubernetes-resources#enable-kubernetes-support to remove this KubernetesConfig class.

As we cannot use ConfigRoot to extract Quarkus properties, do you know how we could collect the kuberntes properties described in 
https://quarkus.io/guides/kubernetes-resources#configuration-options

**Comment by geoand** [comment]
I will defer to @iocanel who is the most qualified person to answer this :)

**Comment by iocanel** [comment]
`KubernetesConfig` has been removed. We need to update the doc.

Currently, all configuration options originating from dekorate objects.

These objects are:

io.dekorate.kubernetes.config.KubernetesConfig
io.dekorate.docker.config.DockerBuildConfig
io.dekorate.openshift.config.OpenshiftConfig
io.dekorate.s2i.config.S2iBuildConfig

On top of that, we also have `kubernetes.deployment.target` which is not part of any Object.

Can we get vscode-quarkus use these objects?

**Comment by angelozerr** [comment]
Thanks @iocanel for your great feedback!

> Can we get vscode-quarkus use these objects?

Indeed it's the idea, introspect each fields from KubernetesConfig (and their extends class like BaseConfig) to generate kubernetes properties.

In short today we collect Quarkus properties by scanning all classes from classpath (and deployment JAR) which are annotated with @ConfigRoot and ConfigProperty. It seems Kubernetes extension doesn't use this mechanism.

The question is how to know that kubernetes properties must collect fields from KubernetesConfig ? My fear is to do that by hard coding this logic.

When pom.xml contains:

```xml
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-kubernetes</artifactId>
    </dependency>
```

We hard code the following process:

 * load `quarkus-kubernetes-deployment` in the classpath to load `kubernetes-annotations` which provides KubernetesConfig  class.
 * hard code the get of KubernetesConfig .
 * create properties by looping each fields from KubernetesConfig class (and their base class).

Let me explain more the idea:

 * KubernetesConfig which extends BaseConfig:

```java
package io.dekorate.kubernetes.config;

import java.lang.String;
import io.dekorate.project.Project;
import io.dekorate.kubernetes.annotation.ImagePullPolicy;
import io.dekorate.kubernetes.annotation.ServiceType;
import java.lang.Object;
import java.util.Map;
import io.sundr.builder.annotations.Buildable;

@Buildable(builderPackage = "io.dekorate.deps.kubernetes.api.builder") public class KubernetesConfig extends BaseConfig{

    private io.dekorate.kubernetes.config.Container[] initContainers = new io.dekorate.kubernetes.config.Container[0];
    private int replicas = 1;

    public KubernetesConfig(){
    }
    public KubernetesConfig(Project project,Map<ConfigKey,Object> attributes,String group,String name,String version,io.dekorate.kubernetes.config.Label[] labels,io.dekorate.kubernetes.config.Annotation[] annotations,io.dekorate.kubernetes.config.Env[] envVars,String workingDir,String[] command,String[] arguments,String serviceAccount,String host,io.dekorate.kubernetes.config.Port[] ports,ServiceType serviceType,io.dekorate.kubernetes.config.PersistentVolumeClaimVolume[] pvcVolumes,io.dekorate.kubernetes.config.SecretVolume[] secretVolumes,io.dekorate.kubernetes.config.ConfigMapVolume[] configMapVolumes,io.dekorate.kubernetes.config.GitRepoVolume[] gitRepoVolumes,io.dekorate.kubernetes.config.AwsElasticBlockStoreVolume[] awsElasticBlockStoreVolumes,io.dekorate.kubernetes.config.AzureDiskVolume[] azureDiskVolumes,io.dekorate.kubernetes.config.AzureFileVolume[] azureFileVolumes,io.dekorate.kubernetes.config.Mount[] mounts,ImagePullPolicy imagePullPolicy,String[] imagePullSecrets,io.dekorate.kubernetes.config.Probe livenessProbe,io.dekorate.kubernetes.config.Probe readinessProbe,io.dekorate.kubernetes.config.Container[] sidecars,boolean expose,boolean autoDeployEnabled,io.dekorate.kubernetes.config.Container[] initContainers,int replicas){
            super(project, attributes, group, name, version, labels, annotations, envVars, workingDir, command, arguments, serviceAccount, host, ports, serviceType, pvcVolumes, secretVolumes, configMapVolumes, gitRepoVolumes, awsElasticBlockStoreVolumes, azureDiskVolumes, azureFileVolumes, mounts, imagePullPolicy, imagePullSecrets, livenessProbe, readinessProbe, sidecars, expose, autoDeployEnabled);
            this.initContainers = initContainers != null ? initContainers : new io.dekorate.kubernetes.config.Container[0];
            this.replicas = replicas;
    }

    public io.dekorate.kubernetes.config.Container[] getInitContainers(){
            return this.initContainers;
    }

    public int getReplicas(){
            return this.replicas;
    }
}
```

 * and  BaseConfig looks like:

```java
package io.dekorate.kubernetes.config;

import java.lang.String;
import io.dekorate.project.Project;
import io.dekorate.kubernetes.annotation.ImagePullPolicy;
import io.dekorate.kubernetes.annotation.ServiceType;
import java.lang.Object;
import java.util.Map;
import io.sundr.builder.annotations.Buildable;

@Buildable(builderPackage = "io.dekorate.deps.kubernetes.api.builder") public class BaseConfig extends ApplicationConfiguration{

    private io.dekorate.kubernetes.config.Label[] labels = new io.dekorate.kubernetes.config.Label[0];
    private io.dekorate.kubernetes.config.Annotation[] annotations = new io.dekorate.kubernetes.config.Annotation[0];
    private io.dekorate.kubernetes.config.Env[] envVars = new io.dekorate.kubernetes.config.Env[0];
    private String workingDir = "";
    private String[] command = new java.lang.String[0];
    private String[] arguments = new java.lang.String[0];
    private String serviceAccount = "";
    private String host = "";
    private io.dekorate.kubernetes.config.Port[] ports = new io.dekorate.kubernetes.config.Port[0];
    private ServiceType serviceType = io.dekorate.kubernetes.annotation.ServiceType.ClusterIP;
    private io.dekorate.kubernetes.config.PersistentVolumeClaimVolume[] pvcVolumes = new io.dekorate.kubernetes.config.PersistentVolumeClaimVolume[0];
    private io.dekorate.kubernetes.config.SecretVolume[] secretVolumes = new io.dekorate.kubernetes.config.SecretVolume[0];
    private io.dekorate.kubernetes.config.ConfigMapVolume[] configMapVolumes = new io.dekorate.kubernetes.config.ConfigMapVolume[0];
    private io.dekorate.kubernetes.config.GitRepoVolume[] gitRepoVolumes = new io.dekorate.kubernetes.config.GitRepoVolume[0];
    private io.dekorate.kubernetes.config.AwsElasticBlockStoreVolume[] awsElasticBlockStoreVolumes = new io.dekorate.kubernetes.config.AwsElasticBlockStoreVolume[0];
    private io.dekorate.kubernetes.config.AzureDiskVolume[] azureDiskVolumes = new io.dekorate.kubernetes.config.AzureDiskVolume[0];
    private io.dekorate.kubernetes.config.AzureFileVolume[] azureFileVolumes = new io.dekorate.kubernetes.config.AzureFileVolume[0];
    private io.dekorate.kubernetes.config.Mount[] mounts = new io.dekorate.kubernetes.config.Mount[0];
    private ImagePullPolicy imagePullPolicy = io.dekorate.kubernetes.annotation.ImagePullPolicy.IfNotPresent;
    private String[] imagePullSecrets = new java.lang.String[0];
    private io.dekorate.kubernetes.config.Probe livenessProbe = null;
    private io.dekorate.kubernetes.config.Probe readinessProbe = null;
    private io.dekorate.kubernetes.config.Container[] sidecars = new io.dekorate.kubernetes.config.Container[0];
    private boolean expose = false;
    private boolean autoDeployEnabled = false;
...
```
 * and ApplicationConfiguration which extends BaseConfig looks like:

```java
package io.dekorate.kubernetes.config;

import io.dekorate.Coordinates;
import io.dekorate.project.Project;
import io.sundr.builder.annotations.Buildable;

import java.util.Map;

@Buildable(builderPackage = "io.dekorate.deps.kubernetes.api.builder")
public class ApplicationConfiguration extends Configuration implements Coordinates {

  private String group;
  private String name;
  private String version;

```
With introspection of those classes, we could generate:

 * `kubernetes.replicas` (from KubernetesConfig)
 * `kubernetes.labels[]` (from BaseConfig)
 * `kubernetes.group` (from ApplicationConfiguration)

The question is how to know that we must collect fields from the Java class `io.dekorate.kubernetes.config.KubernetesConfig` with just pom dependency declaration:

```xml
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-kubernetes</artifactId>
    </dependency>
```

JDT Quarkus LS extension is able to load classes from `quarkus-kubernetes-deployment` (without declaring it in pom.xml) which provides [KubernetesProcessor](https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java). After reading the code of this processor I suppose it's this class which load kubernetes properties. But there are no direct link to `io.dekorate.kubernetes.config.KubernetesConfig`. 

My fear is that it uses introspection to load kubernetes properties by `io.dekorate.Session` in https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java#L101

And in this case we have no way to know that we must instropect fields from KubernetesConfig.

**Comment by iocanel** [comment]
What can we do to help?

Would it be of any use if we added a class inside the quarkus kubernetes extension, that lists all root configuration classes?

**Comment by iocanel** [comment]
Another idea is to generate a descriptor in the format of your choice that returns the property names (and type), similarly to how we generate the documentation tables (see https://quarkus.io/guides/kubernetes-resources).

**Comment by angelozerr** [comment]
> Another idea is to generate a descriptor in the format of your choice

We already discussed about that in https://github.com/quarkusio/quarkus/issues/1655 but nothing was done, that's why we collect Quarkus properties by using annotations. IMHO I think we should:

 * have for each Quarkus extension a JSON file which provides all Quarkus properties. It will improve performances I think for collecting Quarkus properties from JAR.
 * collecting Quarkus properties by scanning source to collect properties from ConfigRoot and ConfigProperty annotations (like it is done today).

Here the JSON format that we use after collecting Quarkus properties: https://github.com/redhat-developer/quarkus-ls/blob/master/quarkus.ls/com.redhat.quarkus.ls/src/test/resources/com/redhat/quarkus/services/all-quarkus-properties.json

Here a sample:

```json
{
			"propertyName": "quarkus.datasource.driver",
			"type": "java.util.Optional\u003cjava.lang.String\u003e",
			"docs": "The datasource driver class name",
			"extensionName": "quarkus-agroal",
			"location": "C:/Users/azerr/.m2/repository/io/quarkus/quarkus-agroal/0.21.1/quarkus-agroal-0.21.1.jar",
			"source": "io.quarkus.agroal.runtime.DataSourceBuildTimeConfig#driver",
			"required": false,
			"phase": 2
		}
```

location information could be omitted.

In other words if we have list of properties with name, source (we need that to manage go to definition), it should be very fantastic.

BUT even if we have this feature, I think we should develop specific code to manage the `kubernetes.deployment.target=kubernetes, openshift`

If I understand the documentation, it means that if we declare this property in application.properties, I should have completion for kubernetes.* and openshift*. properties.

It seems that you are opened to generate a descriptor, which is a cool news for vscode-quarkus, the question is. Is this descriptor will be generic and it should work for any Quarkus extensions or just for the kubernetes extension?

 * generic descriptor means that we must write a descriptor which exposes the properties according a condition (if I declare `kubernetes.deployment.target=kubernetes`, I must just have kubernetes.* properties)
 * not generic, it means that you could generate 2 descriptors for kubernetes and openshifht and in Quarkus LS side (vscode-quarkus), we hard code the condition.

@fbricon what do you think about that?

**Comment by fbricon** [comment]
Can't the kubernetes extension use ConfigRoot and ConfigProperty annotations like the rest? That'd make the Quarkus ecosystem more consistent and would then work OOTB in vscode-quarkus

**Comment by maxandersen** [comment]
@angelozerr please open separate issue on your ideas for additional metadata that makes sense to have in extensions - we can definitely look at that. #1655 is done now ;)

**Comment by iocanel** [comment]
Would it help if I provided a json descriptor with these information?

**Comment by angelozerr** [comment]
> Would it help if I provided a json descriptor with these information?

Totally! Is there any chance that you will provide this feature? 

If the answer is not, @fbricon I think we should support it by loading with hard coded the `io.dekorate.kubernetes.config.KubernetesConfig `class to introspect. We could do that in an Eclipse extension point to manage contribution of annotation to scan (like Mircroprofile annotation) or class to load to get Quarkus properties (like KubernetesConfig ).

**Comment by maxandersen** [comment]
@angelozerr i think we've said multiple times now we would like to add metadata approach for those plugins where the static introspection doesn't work well for you. But we aren't going to add a json structure as you suggested which has a lot of internal IDE needs.

Could we outline on  https://github.com/quarkusio/quarkus/issues/5229 the minimal set of data you think you need and we go from there ?

**Comment by angelozerr** [comment]
> @angelozerr i think we've said multiple times now we would like to add metadata approach for those plugins where the static introspection doesn't work well for you. But we aren't going to add a json structure as you suggested which has a lot of internal IDE needs.

Sorry @maxandersen it was not clear for me, no it's very clear! thanks for the clarification!

> Could we outline on quarkusio/quarkus#5229 the minimal set of data you think you need and we go from there ?

done in https://github.com/quarkusio/quarkus/issues/5229#issuecomment-556072590

**Comment by iocanel** [comment]
@maxandersen @angelozerr: I spent some time today trying to generate the config classes that we need (pretty much replicate the dekorate config classes but with @ConfigRoot and @ConfigItem annotations).

One issue I am hitting is that it seems that when I do so, quarkus expects these properties prefixed with `qurakus.` while for the particular extension this is not the case (it has been decided that the properties won't be prefixed with `quarkus.` but with `kubernetes.`, `openshift.` and so on. 

This seems to be a similar case with some of the microprofile properties which are prefixed with `mp.`. How does vscode work with those properties?

**Comment by angelozerr** [comment]
> This seems to be a similar case with some of the microprofile properties which are prefixed with mp.. How does vscode work with those properties?

We support only quarkus properties. We would like to support microprofile properties, have you some documentation about mp prefix please ?

I'm refactoring our Quarkus properties collector to support microprofile, and I would like to support kubernetes., openshift prefix too. 

The question is what is the rule to compute the proper prefix quarkus, mp, kuberntes, openshift ?

**Comment by iocanel** [comment]
> The question is what is the rule to compute the proper prefix quarkus, mp, kuberntes, openshift ?

Excellent question. @maxandersen any idea?

**Comment by maxandersen** [comment]
You mean in the quarkus-kubernetes extension ? then I don't know as it was made before I got around; but I can imagine its just that settings specific to kubernetes and openshift and thus not generic are put "outside" quarkus namespace.

@geoand in addition to @iocanel made most of the contributions to the extension ... I was hoping you would know ;)

**Comment by geoand** [comment]
I'll need to read the discussions here and get back to you :)

**Comment by geoand** [comment]
So here is the current Quarkus situation:

Almost all of the extensions use properties that are prefixed with `quarkus`. A few special cases like the Kubernetes extension and the smallrye reactive streams extensions use other prefixes that are entirely extension specific and there isn't a way for a tool to determine those prefixes currently.
From my perspective, I think that currently just hardcoding the exceptional prefixes for these extensions would be fine.

**Comment by maxandersen** [comment]
@iocanel @geoand as just suggested on https://github.com/quarkusio/quarkus/issues/5229#issuecomment-560306593 we might want to use something like springs config format as that seem to capture most we need .

**Comment by angelozerr** [comment]
@iocanel by waiting for Quarkus metadata, I have started to support kubernetes, openshift properties.

Here a demo:

![KubernetesDemo](https://user-images.githubusercontent.com/1932211/71182413-57a2e500-2276-11ea-8f40-6a883e2fc849.gif)

I'm using Spring Boot format, and here some example:

 * for `kubernetes.name`:

```json
{
	"type": "java.lang.String",
	"sourceField": "name",
	"required": true,
	"phase": 0,
	"name": "kubernetes.name",
	"sourceType": "io.dekorate.kubernetes.config.ApplicationConfiguration"
}
```

 * for `kubernetes.init-containers[*].image` (I'm using [*] for syntax for index array:

```json
{
	"type": "java.lang.String",
	"sourceField": "image",
	"required": true,
	"phase": 0,
	"name": "kubernetes.init-containers[*].image",
	"sourceType": "io.dekorate.kubernetes.config.Container"
}
```

To do that I'm introspecting each fields from the `KubernetesConfig` class.

The thing that I don't understand is about Docker https://quarkus.io/guides/kubernetes#docker. I think there is some bugs with table generation (for instance auto-push-enabled should be on property column).

The question is what is the proper property? For instance the doc says that `kubernetes.docker-file` is a property. KubernetesConfig doesn't provides this dockerFile field, but `DockerBuildConfig` provides it. My feeling is that we must instrospect too `DockerBuildConfig` which `kubernetes` prefix. But if I do that I will have another properties like `kubernetes.auto-push-enabled`, is it a valid property ?

Many thanks for your answer.

**Comment by iocanel** [comment]
Looks awesome!

Properties found in DockerBuildConfig are meant to be prefixed with docker.

**Comment by angelozerr** [comment]
> Looks awesome!

Glad it pleases you :)

> Properties found in DockerBuildConfig are meant to be prefixed with docker.

I have generated properties by following this strategy. 

But I don't understand where `kubernetes.docker-file` property comes from?

I understand that for `docker.docker-file` comes from the ImageConfiguration#dockerFile field (because DockerBuildConfig extends ImageConfiguration) but KubernetesConfig doesn't extend ImageConfiguration, so my question is how to generate `kubernetes.docker-file` property ?

**Comment by angelozerr** [comment]
@iocanel I wonder if it should be better to use https://github.com/dekorateio/dekorate/blob/master/annotations/kubernetes-annotations/src/main/java/io/dekorate/kubernetes/annotation/KubernetesApplication.java instead of generated class KubernetesConfig to generate properties. 

Using KubernetesApplication will provide:

 * the documentation extract from Javadoc
 * the default value

What do you think about that? 

The thing is that I don't understand is that you duplicate some fields in KubernetesApplication  (which generates KubernetesConfig class which extends BaseClass) and Base (which generates BaseConfig class) annotation, why KubernetesApplication  doesn't extend Base annotation?

**Comment by iocanel** [comment]
AFAIK it's not possible one annotation to extend another.

**Comment by angelozerr** [comment]
Forger m'y commentés you are right.

But could you tell me if it should be a better Idea to use
KUbernetesApplication instead of kubernetesconfig class

Le sam. 21 déc. 2019 18:08, Ioannis Canellos <notifications@github.com> a
écrit :

**Comment by iocanel** [comment]
There shouldn't be much of a difference as the config is generated from the annotation. Both should work fine. The annotation has the benefit of being flat (no inherited fields).

**Comment by fbricon** [comment]
fixed with https://github.com/redhat-developer/quarkus-ls/pull/170

**Comment by xorye** [comment]
Hi @iocanel, is it normal for the Quarkus Kubernetes deployment JAR to not have a dependency to s2i? If this is normal, which dependency can I add to the pom to have the s2i build annotation?

**Comment by iocanel** [comment]
You shouldn't need any additional jar. If you configure open shift as the target platform s2i related resources will be created for you (image streams and build config).

**Comment by angelozerr** [comment]
@iocanel the kubernetes ans openshift properties are generated from the kubernetesapplication ans openshift application annotations comming from the dekorate jar. This dekorate jar is added in the classpath when you declare thé quarkus-kubernetes depency in your pom.xml. in this case WE load the quarkus-kubernes-deployment jar which have dependencies in his pom to kubernetes and openshift dekorate jars but there are no dependency to s2i dekorate annotation? Perhaps the quarkus-kubernetes-dependency should have a dependency to the s2i dekorate jar,no?

**Comment by iocanel** [comment]
It's a transitive. openshift-annotations depends on s2i-annotations.

**Comment by angelozerr** [comment]
@iocanel thanks a lot for your answer. It was a bug on our side that I have tried to fix it in https://github.com/redhat-developer/quarkus-ls/pull/179 

![image](https://user-images.githubusercontent.com/1932211/71769927-1c1b4280-2f27-11ea-9619-a4d420353ff4.png)

@xorye could you review it please?

**Comment by iocanel** [comment]
Looks awesome!
